### DISCUSSION TITLE
The Anatomy of ct.sym — How javac Ensures Backwards Compatibility - Gunnar Morling

### DISCUSSION THREAD

**Discussion by blog-comment-importer** [gunnarmorling/discussions.morling.dev#283] [Announcements]
https://www.morling.dev/blog/the-anatomy-of-ct-sym-how-javac-ensures-backwards-compatibility/

**Comment by blog-comment-importer** [comment]
_**Alexander Kriegisch** wrote at 2021-06-19T03:54:30Z (comment imported from Disqus):_

<p>Thanks for this nice summary. I found out most of it by myself, re-inventing the wheel you had already designed here. I wish I would have found it earlier. Do you have any links explaining how <code>ct.sym</code> is generated and if the same tool could be used for my own purposes?</p><p>Background: In an OSS project I am contributing to, there is kind of the <b>reverse case:</b> Code is compiled against a <b>more recent</b> JDK API, but compilation should also work on older JDKs, which means the project manually creates skeleton classes for the more recent API, using it on the older JDK. Maybe the same JDK tool would come in handy there.</p>

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2021-06-19T08:06:40Z (comment imported from Disqus):_

<p>Thanks for the feedback; happy to hear you found the post useful!</p><p></p><blockquote>Do you have any links explaining how is generated and if the same tool could be used for my own purposes?</blockquote><p></p><p>You could have a look at <a href="https://github.com/openjdk/jdk/blob/master/make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java" rel="nofollow noopener" title="https://github.com/openjdk/jdk/blob/master/make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java">CreateSymbols.java</a> in the JDK. I don't think it does exactly what you want but it may serve as basis for inspiration.</p>

**Reply by blog-comment-importer** [reply]
_**Alexander Kriegisch** wrote at 2021-06-20T01:17:04Z (comment imported from Disqus):_

<p>Thanks for the precious hint, Gunnar. 🙂</p>

**Comment by astubbs** [comment]
What I want to know is why Oracle doesn't expose the parsers of this file outside their module. It is quite useful for a lot of things..
